### PR TITLE
MDS-324 | Add stub V2 endpoints

### PR DESCRIPTION
**Pull request by owen-opencast**

### CODE REVIEW

File: test/component/uk/gov/hmrc/individualsincomeapi/controllers/v2/IndividualIncomeSpec.scala [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+    scenario("Individual has employment income") {
+
+      Given("A valid privileged Auth bearer token")
+      AuthStub.willAuthorizePrivilegedAuthToken(authToken, payeIncomeScope, retrieveAll = true)
+
+      And("a valid record in the matching API")
+      IndividualsMatchingApiStub.hasMatchFor(matchId, nino)
+
+      And("DES will return employments for the NINO")
```
Comment (ajchilts) [inline test/component/uk/gov/hmrc/individualsincomeapi/controllers/v2/IndividualIncomeSpec.scala:47]: V2 will be on IF and not DES. Can we please remove the references to DES?

File: test/component/uk/gov/hmrc/individualsincomeapi/controllers/v2/LiveSaIncomeControllerSpec.scala [hunk 2]
```diff
@@ -0,0 +1,568 @@
...
+import uk.gov.hmrc.individualsincomeapi.domain.{DesSAIncome, DesSAReturn, SAIncome, TaxYear}
+
+class LiveSaIncomeControllerSpec extends BaseSpec {
+
+  val matchId = UUID.randomUUID().toString
+  val nino = Nino("AA100009B")
+  val fromTaxYear = TaxYear("2016-17")
+  val toTaxYear = TaxYear("2018-19")
+  val desIncomes = Seq(
+    DesSAIncome(
```
Comment (ajchilts) [inline test/component/uk/gov/hmrc/individualsincomeapi/controllers/v2/LiveSaIncomeControllerSpec.scala:36]: V2 controller integrates with IF not DES. Can we remove the references to DES as they should not be needed for not yet implemented?

File: test/component/uk/gov/hmrc/individualsincomeapi/controllers/v2/LiveSaIncomeControllerSpec.scala [hunk 3]
```diff
@@ -0,0 +1,568 @@
...
+      Given("A privileged Auth bearer token with scope read:individuals-income-sa-interests-and-dividends")
+      AuthStub.willAuthorizePrivilegedAuthToken(
+        authToken,
+        "read:individuals-income-sa-interests-and-dividends",
+        retrieveAll = true)
+
+      And("a valid record in the matching API")
+      IndividualsMatchingApiStub.willRespondWith(matchId, OK, s"""{"matchId" : "$matchId", "nino" : "$nino"}""")
+
+      And("DES will return self-assessment data for the individual")
```
Comment (ajchilts) [inline test/component/uk/gov/hmrc/individualsincomeapi/controllers/v2/LiveSaIncomeControllerSpec.scala:367]: Please remove all references to DES.

File: test/unit/uk/gov/hmrc/individualsincomeapi/controllers/v2/LiveRootControllerSpec.scala [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+      when(mockLiveCitizenMatchingService.matchCitizen(refEq(randomMatchId))(any[HeaderCarrier]))
+        .thenReturn(successful(MatchedCitizen(randomMatchId, Nino("AB123456C"))))
+
+      val result = intercept[Exception] { await(liveMatchCitizenController.root(randomMatchId).apply(FakeRequest())) }
+      assert(result.getMessage == "NOT_IMPLEMENTED")
+    }
+
+    "fail with AuthorizedException when the bearer token does not have enrolment read:individuals-income" in new Setup {
+//      given(
+//        mockAuthConnector.authorise(refEq(Enrolment("read:individuals-income")), refEq(EmptyRetrieval))(any(), any()))
```
Comment (ajchilts) [inline test/unit/uk/gov/hmrc/individualsincomeapi/controllers/v2/LiveRootControllerSpec.scala:75]: Can we delete this commented out code?

**Review by ajchilts — CHANGES_REQUESTED** [review]
Great effort 👍 

Few comments.

File: test/component/uk/gov/hmrc/individualsincomeapi/VersioningSpec.scala [hunk 5]
```diff
@@ -54,6 +54,20 @@ class VersioningSpec extends BaseSpec {
...
     }
 
+    scenario("Requests with an accept header version P2.0") {
+      Given("A valid privileged Auth bearer token")
+      AuthStub.willAuthorizePrivilegedAuthToken(authToken, incomeScope)
+
+      When("A request to the match citizen endpoint is made with version P2.0 accept header")
+      val response = invokeWithHeaders(s"/sandbox?matchId=$sandboxMatchId", AUTHORIZATION -> authToken, acceptHeaderP2)
+
+      Then("The response status should be 200")
```
Comment (ajchilts) [inline test/component/uk/gov/hmrc/individualsincomeapi/VersioningSpec.scala:64]: Test specifies 200 then we expect INTERNAL_SERVER_ERROR?

Please correct all occurrences.

**Review by ajchilts — CHANGES_REQUESTED** [review]

File: conf/app_version_P2.routes [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+GET        /sa/pensions-and-state-benefits  uk.gov.hmrc.individualsincomeapi.controllers.v2.LiveSaIncomeController.saPensionsAndStateBenefitsIncome(matchId: java.util.UUID, taxYearInterval: uk.gov.hmrc.individualsincomeapi.domain.TaxYearInterval)
+GET        /sa/uk-properties                uk.gov.hmrc.individualsincomeapi.controllers.v2.LiveSaIncomeController.saUkPropertiesIncome(matchId: java.util.UUID, taxYearInterval: uk.gov.hmrc.individualsincomeapi.domain.TaxYearInterval)
+GET        /sa/additional-information       uk.gov.hmrc.individualsincomeapi.controllers.v2.LiveSaIncomeController.saAdditionalInformation(matchId: java.util.UUID, taxYearInterval: uk.gov.hmrc.individualsincomeapi.domain.TaxYearInterval)
+GET        /sa/other                        uk.gov.hmrc.individualsincomeapi.controllers.v2.LiveSaIncomeController.saOtherIncome(matchId: java.util.UUID, taxYearInterval: uk.gov.hmrc.individualsincomeapi.domain.TaxYearInterval)
+GET        /sa/source                       uk.gov.hmrc.individualsincomeapi.controllers.v2.LiveSaIncomeController.saIncomeSource(matchId: java.util.UUID, taxYearInterval: uk.gov.hmrc.individualsincomeapi.domain.TaxYearInterval)
```
[the remark this answers is not in the document]
Comment (owen-opencast) [reply conf/app_version_P2.routes:16]: Added further-details now, not sure why it's not appearing here? Can someone pull the branch and check?

File: app/uk/gov/hmrc/individualsincomeapi/controllers/v2/SaIncomeController.scala [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+      {
+        val scopes = scopeService.getEndPointScopes("income-sa-self-employments")
+        requiresPrivilegedAuthentication(scopes)
+          .flatMap { authScopes =>
+            throw new Exception("NOT_IMPLEMENTED")
+          }
+          .recover(recovery)
+      }
+  }
+}
```
[the remark this answers is not in the document]
Comment (owen-opencast) [reply app/uk/gov/hmrc/individualsincomeapi/controllers/v2/SaIncomeController.scala:208]: See above

File: conf/sandbox_version_P2.routes [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+GET        /sa/uk-properties                uk.gov.hmrc.individualsincomeapi.controllers.v2.SandboxSaIncomeController.saUkPropertiesIncome(matchId: java.util.UUID, taxYearInterval: uk.gov.hmrc.individualsincomeapi.domain.TaxYearInterval)
+GET        /sa/pensions-and-state-benefits  uk.gov.hmrc.individualsincomeapi.controllers.v2.SandboxSaIncomeController.saPensionsAndStateBenefitsIncome(matchId: java.util.UUID, taxYearInterval: uk.gov.hmrc.individualsincomeapi.domain.TaxYearInterval)
+GET        /sa/additional-information       uk.gov.hmrc.individualsincomeapi.controllers.v2.SandboxSaIncomeController.saAdditionalInformation(matchId: java.util.UUID, taxYearInterval: uk.gov.hmrc.individualsincomeapi.domain.TaxYearInterval)
+GET        /sa/other                        uk.gov.hmrc.individualsincomeapi.controllers.v2.SandboxSaIncomeController.saOtherIncome(matchId: java.util.UUID, taxYearInterval: uk.gov.hmrc.individualsincomeapi.domain.TaxYearInterval)
+GET        /sa/source                       uk.gov.hmrc.individualsincomeapi.controllers.v2.SandboxSaIncomeController.saIncomeSource(matchId: java.util.UUID, taxYearInterval: uk.gov.hmrc.individualsincomeapi.domain.TaxYearInterval)
```
[the remark this answers is not in the document]
Comment (owen-opencast) [reply conf/sandbox_version_P2.routes:17]: See above

**Review by ajchilts — APPROVED** [review]
LGTM 👍

**Comment by mostafasina** [comment]
Regression tests in local dev passed.

**Pull request merged**